src/provider.rs

```diff
@@ -409,14 +418,25 @@ impl Future for Provider {
     }
 }
 
-struct Handler(Database);
+#[derive(Debug, Clone)]
+struct ProviderHandles {
+    /// database handle
+    pub db: Database,
+    /// cancellation token
+    pub shutdown: CancellationToken,
+    /// peer id
+    pub peer_id: PeerId,
+    /// auth token
+    pub auth_token: AuthToken,
+    /// listen address
+    pub listen_addr: SocketAddr,
+}
```

Comment:
I guess the main difference with `Provider` is that you can clone this?  And the main thing that can't be cloned in the `Provider` is the `JoinHandle` afaik.  Maybe this is fine, maybe we can build `Provider` out of the `JoinHandle` & `ProviderHandle`?  And if going that way maybe we can have `Provider::handle()` as a public API at some point.

I guess for now the question of my rambling is: can we make these two more the same and use `ProviderHandles` inside the `Provider` to avoid duplication?

Reply:
Yes, I was almost at the point where I wanted to make Provider cloneable. I did not do it because I saw that provider implements Future.

We could make Provider cloneable. It is already not really "the provider" (that is some tokio task) but some kind of handle that allows you to get info for the provider.

WDYT?

Reply:
How about we see if we can make Provider a proper handle in a subsequent PR?